### PR TITLE
perf: improve i18n caching

### LINKED ISSUE #9840: [CAL-2070] reduce i18n skeletons
we have a lot of skeletons in place where we are loading i18n strings (translated words)

if we dont have these skeletons, it would show something like "event_types_title" for a brief second

if we do SSR for all pages, we'd wait for the string to load and then load the entire page, causing no skeletons, no string flickering but slower pages (we are bascially blocking the UI till all strings are loaded)

\_\_

is there no way to load the entire en/common.json once, and keep it somehow warm/around without constant skeletons?

ideally we get away without SSR

EDIT by @zomars:

Made a more detailed explanation of the desired solution here: [https://github.com/calcom/cal.com/issues/10198](https://github.com/calcom/cal.com/issues/10198)

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-2070](https://linear.app/calcom/issue/CAL-2070/reduce-i18n-skeletons)</sub>

### PR DESCRIPTION
**Pull request by Aditay2002**
fixes #9840 

/claim #9840

### CODE REVIEW

File: apps/web/next.config.js [hunk 1]
```diff
@@ -58,7 +62,20 @@ if (process.env.GOOGLE_API_CREDENTIALS && !validJson(process.env.GOOGLE_API_CRED
   );
 }
 
+const getCachedEnglishTranslation = () => {
+  const cachedTranslation = cache.get('englishTranslation');
```
Comment (PeerRich) [inline apps/web/next.config.js:66]:
```suggestion
  const cachedTranslation = cache.get("englishTranslation");
```

File: apps/web/next.config.js [hunk 2]
```diff
@@ -58,7 +62,20 @@ if (process.env.GOOGLE_API_CREDENTIALS && !validJson(process.env.GOOGLE_API_CRED
...
+  if (cachedTranslation) {
+    return cachedTranslation;
+  } else {
+    const englishTranslation = fetchEnglishTranslation();
+    cache.put('englishTranslation', englishTranslation);
```
Comment (PeerRich) [inline apps/web/next.config.js:71]:
```suggestion
    cache.put("englishTranslation", englishTranslation);
```

File: apps/web/next.config.js [hunk 3]
```diff
@@ -1,7 +1,11 @@
 require("dotenv").config({ path: "../../.env" });
 const CopyWebpackPlugin = require("copy-webpack-plugin");
+const cache = require('memory-cache');
```
Comment (PeerRich) [inline apps/web/next.config.js:3]:
```suggestion
const cache = require("memory-cache");
```

File: apps/web/next.config.js [hunk 4]
```diff
@@ -1,7 +1,11 @@
...
 const os = require("os");
-const englishTranslation = require("./public/static/locales/en/common.json");
+const fetchEnglishTranslation = () => {
+ 
+  return require('./public/static/locales/en/common.json');
```
Comment (PeerRich) [inline apps/web/next.config.js:7]:
```suggestion
  return require("./public/static/locales/en/common.json");
```

File: apps/web/next.config.js [hunk 5]
```diff
@@ -1,7 +1,11 @@
 require("dotenv").config({ path: "../../.env" });
 const CopyWebpackPlugin = require("copy-webpack-plugin");
+const cache = require("memory-cache");
 const os = require("os");
-const englishTranslation = require("./public/static/locales/en/common.json");
+const fetchEnglishTranslation = () => {
+
```
Comment (PeerRich) [inline apps/web/next.config.js:6]:
```suggestion
```

File: apps/web/next.config.js [hunk 6]
```diff
@@ -58,7 +62,20 @@ if (process.env.GOOGLE_API_CREDENTIALS && !validJson(process.env.GOOGLE_API_CRED
...
+  } else {
+    const englishTranslation = fetchEnglishTranslation();
+    cache.put("englishTranslation", englishTranslation);
+    return englishTranslation;
+  }
+};
+
 const informAboutDuplicateTranslations = () => {
+  const englishTranslation = getCachedEnglishTranslation();
+
```
Comment (PeerRich) [inline apps/web/next.config.js:78]:
```suggestion
```

**Comment by PeerRich** [comment]
this seems to be only caching the english translation, no?

**Comment by PeerRich** [comment]
can you make a self-review? https://handbook.cal.com/engineering/self-reviews

File: apps/web/next.config.js [hunk 7]
```diff
@@ -1,7 +1,10 @@
 require("dotenv").config({ path: "../../.env" });
 const CopyWebpackPlugin = require("copy-webpack-plugin");
+const cache = require("memory-cache");
 const os = require("os");
```
Comment (Aditay2002) [inline apps/web/next.config.js:4]: here memory cahce is added to the config.js file

File: apps/web/next.config.js [hunk 8]
```diff
@@ -1,7 +1,10 @@
...
-const englishTranslation = require("./public/static/locales/en/common.json");
```
Comment (Aditay2002) [inline apps/web/next.config.js:4]: englishTranslation is just replaced with a function and remove from here.

File: apps/web/next.config.js [hunk 9]
```diff
@@ -1,7 +1,10 @@
...
+const fetchEnglishTranslation = () => {
+  return require("./public/static/locales/en/common.json");
+};
 const { withAxiom } = require("next-axiom");
```
Comment (Aditay2002) [inline apps/web/next.config.js:8]: here is the function to fetch the englishTranslation when ever required.

File: apps/web/next.config.js [hunk 10]
```diff
@@ -58,7 +61,19 @@ if (process.env.GOOGLE_API_CREDENTIALS && !validJson(process.env.GOOGLE_API_CRED
...
+  const cachedTranslation = cache.get("englishTranslation");
+  if (cachedTranslation) {
+    return cachedTranslation;
+  } else {
+    const englishTranslation = fetchEnglishTranslation();
+    cache.put("englishTranslation", englishTranslation);
+    return englishTranslation;
+  }
+};
+
```
Comment (Aditay2002) [inline apps/web/next.config.js:74]: 1)The function getCachedEnglishTranslation is defined as a constant arrow function, which means it cannot be reassigned.

2)The function first attempts to retrieve the English translation from the cache using the key "englishTranslation" by calling cache.get("englishTranslation"). The result is stored in the cachedTranslation constant.

3)If cachedTranslation is truthy (i.e., the translation exists in the cache), it is returned immediately.

4)If cachedTranslation is falsy (i.e., the translation is not in the cache), the function proceeds to fetch the English translation by invoking the fetchEnglishTranslation function. The result is stored in the englishTranslation constant.

5)The newly fetched englishTranslation is then stored in the cache using the key "englishTranslation" by calling cache.put("englishTranslation", englishTranslation).

6)Finally, the englishTranslation is returned as the result of the function.

File: apps/web/next.config.js [hunk 11]
```diff
@@ -58,7 +61,19 @@ if (process.env.GOOGLE_API_CREDENTIALS && !validJson(process.env.GOOGLE_API_CRED
...
 const informAboutDuplicateTranslations = () => {
+  const englishTranslation = getCachedEnglishTranslation();
```
Comment (Aditay2002) [inline apps/web/next.config.js:76]: here englishTranslation is being stored by calling  getCachedEnglishTranslation function

File: package.json [hunk 12]
```diff
@@ -94,6 +94,7 @@
     "city-timezones": "^1.2.1",
     "eslint": "^8.34.0",
     "lucide-react": "^0.171.0",
+    "memory-cache": "^0.2.0",
```
Comment (Aditay2002) [inline package.json:97]: dependency for memory-cache is added in package.json

**Review by Aditay2002 — COMMENTED** [review]
self-review done.

**Comment by Aditay2002** [comment]
@PeerRich hey, I have done the self-review.

**Comment by Aditay2002** [comment]
> 

yes.

**Review by hariombalhara — CHANGES_REQUESTED** [review]
This doesn't cache the i18n data, it is doing caching in the duplicates detection flow when we build the app.

We need to be able to cache this
Screenshot 2023-06-29 at 6 14 43 PM

**Comment by Aditay2002** [comment]
@hariombalhara @PeerRich  can you please explain more about the change you needed to be done.

**Comment by hariombalhara** [comment]
@Aditay2002  The request I shared is used throughout the app and is constantly fetched on every route change. Till the response is received, skeletons are shown. So, if we are able to cache that request in browser, then on route change skeleton loading shouldn't happen and it would be a huge improvement.

File: packages/app-store/routing-forms/pages/forms/[...appPages].tsx [hunk 13]
```diff
@@ -109,7 +111,11 @@ export default function RoutingForms({
       description: t("download_responses_description"),
     },
   ];
-
+ const validateFormName = (name: string): void => {
+    if (isFormNameEmpty(name)) {
+      throw new Error("Form name is required");
+    }
+  };
```
Comment (ujwalkumar1995) [inline packages/app-store/routing-forms/pages/forms/[...appPages].tsx:118]: Hey @Aditay2002 
Is this fix for issue #9871?
Could you please share screenshots or some video for this?
I was also looking into the same issue and I am not sure how this fixes the problem.

**Pull request closed without merging**